WalletWasabi/Blockchain/Transactions/SmartTransaction.cs

```diff
@@ -106,8 +139,143 @@ public SmartTransaction(Transaction transaction, Height height, uint256? blockHa
 	/// </summary>
 	public bool IsRBF => !Confirmed && (Transaction.RBF || IsReplacement || WalletInputs.Any(x => x.IsReplaceable()));
 
+	public IReadOnlyCollection<IndexedTxIn> ForeignInputs
+	{
+		get
+		{
+			if (recomputeForeignInputs)
+			{
+				_foreignInputs = GetForeignInputs();
+				recomputeForeignInputs = false;
+			}
+			return _foreignInputs;
+		}
+	}
+
+	public IReadOnlyCollection<IndexedTxOut> ForeignOutputs
+	{
+		get
+		{
+			if (recomputeForeignOutputs)
+			{
+				_foreignOutputs = GetForeignOutputs();
+				recomputeForeignOutputs = false;
+			}
+			return _foreignOutputs;
+		}
+	}
+
+	public IReadOnlyCollection<WalletVirtualInput> WalletVirtualInputs
+	{
+		get
+		{
+			if (recomputeWalletVirtualInputs)
+			{
+				_walletVirtualInputs = GetWalletVirtualInputs();
+				recomputeWalletVirtualInputs = false;
+			}
+			return _walletVirtualInputs;
+		}
+	}
+
+	public IReadOnlyCollection<WalletVirtualOutput> WalletVirtualOutputs
+	{
+		get
+		{
+			if (recomputeWalletVirtualOutputs)
+			{
+				_walletVirtualOutputs = GetWalletVirtualOutputs();
+				recomputeWalletVirtualOutputs = false;
+			}
+			return _walletVirtualOutputs;
+		}
+
+	}
+
+	public IReadOnlyCollection<ForeignVirtualOutput> ForeignVirtualOutputs
+	{
+		get
+		{
+			if (recomputeForeignVirtualOutputs)
+			{
+				_foreignVirtualOutputs = GetForeignVirtualOutputs();
+				recomputeForeignVirtualOutputs = false;
+			}
+			return _foreignVirtualOutputs;
+		}
+	}
+
+	private HashSet<IndexedTxIn> _foreignInputs;
+	private HashSet<IndexedTxOut> _foreignOutputs;
+
+	/// <summary>
+	/// Coins those are on the input side of the tx and belong to ANY loaded wallet. Later if more wallets are loaded this list can increase.
+	/// </summary>
+	private HashSet<SmartCoin> _walletInputs;
+
+	/// <summary>
+	/// Coins those are on the output side of the tx and belong to ANY loaded wallet. Later if more wallets are loaded this list can increase.
+	/// </summary>
+	private HashSet<SmartCoin> _walletOutputs;
+
+	private HashSet<WalletVirtualInput> _walletVirtualInputs;
+	private HashSet<WalletVirtualOutput> _walletVirtualOutputs;
+	private HashSet<ForeignVirtualOutput> _foreignVirtualOutputs;
+
+	private bool recomputeForeignInputs = true;
+	private bool recomputeForeignOutputs = true;
+
+	private bool recomputeWalletVirtualInputs = true;
+	private bool recomputeWalletVirtualOutputs = true;
+	private bool recomputeForeignVirtualOutputs = true;
```

Comment:
Rather than having these recompute flags, wouldn't it be possible to assign `null` to the corresponding variable when it needs to be recomputed? Or C# doesn't allow that?

Reply:
Definitely possible and better than my approach.

Done in https://github.com/zkSNACKs/WalletWasabi/pull/7853/commits/35ae3a5840462806726fe9a4c1e7b7897c98ec33 and https://github.com/zkSNACKs/WalletWasabi/pull/7853/commits/3333318d24c80a0f2fc18f87e806594146541a0d.